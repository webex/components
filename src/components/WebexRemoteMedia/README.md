# Webex Remote Media Component

Webex remote media component displays the remote stream of a meeting.

<p align="center">
  <img src="./WebexRemoteMedia.png" alt="Default Webex Remote Media" />
</p>

## Preview

To see all the different possible states of the Webex Remote Media component, you can run our Storybook:

```shell
  npm start
```

## Embed

1.  Create a component adapter from which the data will be retrieved (See [adapters](../../adapters)). For instance:

    ```js
    const jsonAdapter = new WebexJSONAdapter(jsonData);
    ```

2.  Create a component instance by passing the person ID as a string and
    enclose it within [a data provider](../WebexDataProvider/WebexDataProvider.js)
    that takes the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously

    ```js
    <WebexDataProvider adapter={jsonAdapter}>
      <WebexRemoteMedia meetingID="meetingID" />
    </WebexDataProvider>
    ```

The component knows how to manage its data. If anything changes in the data source that the adapter manages, the component will also update on its own.
