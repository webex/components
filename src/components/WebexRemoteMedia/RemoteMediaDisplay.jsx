/* eslint-disable jsx-a11y/media-has-caption, react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';

import {useStream} from '../hooks';

/**
 * Remote Media Display component displays the meeting's remote video and audio html elements
 *
 * @param {object} props
 * @param {object} props.remoteAudio
 * @param {object} props.remoteShare
 * @param {object} props.remoteVideo
 * @returns {object} JSX of the component
 */
export default function RemoteMediaDisplay({remoteAudio, remoteShare, remoteVideo}) {
  const audioRef = useStream(remoteAudio);
  const videoRef = useStream(remoteVideo);
  const shareRef = useStream(remoteShare);

  return (
    <>
      {remoteVideo ? <video ref={videoRef} playsInline autoPlay /> : null}

      {remoteShare ? <video ref={shareRef} playsInline autoPlay /> : null}

      {remoteAudio ? <audio ref={audioRef} autoPlay /> : null}
    </>
  );
}

RemoteMediaDisplay.propTypes = {
  remoteAudio: PropTypes.object,
  remoteShare: PropTypes.object,
  remoteVideo: PropTypes.object,
};

RemoteMediaDisplay.defaultProps = {
  remoteAudio: null,
  remoteShare: null,
  remoteVideo: null,
};
