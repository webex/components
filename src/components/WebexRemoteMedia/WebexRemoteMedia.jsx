/* eslint-disable jsx-a11y/media-has-caption */
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Badge, Spinner, AlertBanner} from '@momentum-ui/react';

import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {TABLET, DESKTOP, DESKTOP_LARGE} from '../breakpoints';
import {
  useElementDimensions,
  useMeeting,
  useMembers,
  useStream,
} from '../hooks';

/**
 * Webex Remote Media component displays the meeting's remote video
 *
 * @param {object} props
 * @param {string} props.className
 * @param {string} props.meetingID
 * @returns {object} JSX of the component
 *
 * NOTE: waiting for the UX for a design on what to display if there is no remote video
 */
export default function WebexRemoteMedia({className, meetingID}) {
  const {
    remoteAudio,
    remoteVideo,
    remoteShare,
    error,
  } = useMeeting(meetingID);
  const members = useMembers(meetingID, 'meeting');
  const audioRef = useStream(remoteAudio);
  const videoRef = useStream(remoteVideo);
  const shareRef = useStream(remoteShare);
  const [remoteMediaRef, {width}] = useElementDimensions();

  const hasOtherMembers = members.length > 1;
  const hasMedia = !!(remoteAudio || remoteVideo || remoteShare);

  const classBaseName = `${WEBEX_COMPONENTS_CLASS_PREFIX}-remote-media`;
  const mainClasses = {
    [`${classBaseName}`]: true,
    [`${classBaseName}-tablet`]: width >= TABLET && width < DESKTOP,
    [`${classBaseName}-desktop`]: width >= DESKTOP && width < DESKTOP_LARGE,
    [`${classBaseName}-desktop-xl`]: width >= DESKTOP_LARGE,
    [className]: !!className,
    'remote-video-n-share': remoteVideo && remoteShare,
  };

  let remoteDisplay;

  if (error) {
    remoteDisplay = (
      <AlertBanner show type="warning">
        Having trouble joining the meeting? Please check your connection.
      </AlertBanner>
    );
  } else if (hasMedia && hasOtherMembers) {
    remoteDisplay = (
      <>
        {remoteVideo
          && <video className={`${WEBEX_COMPONENTS_CLASS_PREFIX}-remote-video`} ref={videoRef} muted playsInline autoPlay />}
        {remoteShare
          && <video className={`${WEBEX_COMPONENTS_CLASS_PREFIX}-remote-share`} ref={shareRef} muted playsInline autoPlay />}
        {remoteAudio && <audio ref={audioRef} autoPlay />}
      </>
    );
  } else if (hasMedia && !hasOtherMembers) {
    remoteDisplay = (
      <h4>
        Waiting for others to join...
      </h4>
    );
  } else {
    remoteDisplay = (
      <Badge rounded>
        <Spinner size={16} />
        <div>Connecting</div>
      </Badge>
    );
  }

  return (
    <div ref={remoteMediaRef} className={classNames(mainClasses)}>
      {remoteDisplay}
    </div>
  );
}

WebexRemoteMedia.propTypes = {
  className: PropTypes.string,
  meetingID: PropTypes.string.isRequired,
};

WebexRemoteMedia.defaultProps = {
  className: '',
};
