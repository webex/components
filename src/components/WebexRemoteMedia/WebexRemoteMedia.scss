.#{$WEBEX_COMPONENTS_CLASS_PREFIX}-remote-media {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  width: 100%;
  position: relative;
  background: white;

  .md-alert-banner {
    position: absolute;
    top: 0;
    width: 100%;
  }

  .md-badge {
    margin: auto;
    padding: 0.4rem;

    div {
      margin-left: 0.4rem;
    }
  }

  video {
    height: 100%;
    width: 100%;
  }

  &.split-screen{
    video {
      width: 50%;
      height: auto;
    }
  }
}
