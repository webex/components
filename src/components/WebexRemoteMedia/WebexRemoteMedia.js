import React from 'react';
import PropTypes from 'prop-types';

import {useMeeting, useStream} from '../hooks';

import './WebexRemoteMedia.scss';

/**
 * Webex Remote Media component displays the meeting's remote video
 *
 * @param {object} props
 * @returns {object} JSX of the component
 *
 * NOTE: waiting for the UX for a design on what to display if there is no remote video
 */
export default function WebexRemoteMedia({meetingID}) {
  const {remoteAudio, remoteVideo} = useMeeting(meetingID);
  const audioRef = useStream(remoteAudio);
  const videoRef = useStream(remoteVideo);

  return (
    <div className="remote-media">
      {remoteVideo ? <video ref={videoRef} playsInline /> : null}
      {remoteAudio ? <audio ref={audioRef} /> : null}
    </div>
  );
}

WebexRemoteMedia.propTypes = {
  meetingID: PropTypes.string.isRequired,
};
