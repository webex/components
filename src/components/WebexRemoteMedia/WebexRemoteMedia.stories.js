import React from 'react';
import WebexRemoteMedia from './WebexRemoteMedia';

export default {
  title: 'Meetings/Webex Remote Media',
  component: WebexRemoteMedia,
};

const Template = (args) => <WebexRemoteMedia {...args} />;

export const Loading = Template.bind({});
Loading.args = {
  meetingID: 'noMedia',
};

export const WaitingForOthers = Template.bind({});
WaitingForOthers.args = {
  meetingID: 'noMembers',
};

export const RemoteAudioOnly = Template.bind({});
RemoteAudioOnly.args = {
  meetingID: 'remoteAudio',
};

export const RemoteVideoOnly = Template.bind({});
RemoteVideoOnly.args = {
  meetingID: 'remoteVideo',
};

export const AllRemoteMedia = Template.bind({});
AllRemoteMedia.args = {
  meetingID: 'remoteAudio&Video',
};

export const Error = Template.bind({});
Error.args = {
  meetingID: 'failMeetingID',
};
