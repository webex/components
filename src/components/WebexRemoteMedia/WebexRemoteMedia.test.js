import React from 'react';

import WebexRemoteMedia from './WebexRemoteMedia';

jest.mock('../hooks/useMeeting');
jest.mock('../hooks/useStream');

describe('Webex Remote Media component', () => {
  describe('snapshot', () => {
    test('matches snapshot while loading', () => {
      expect(shallow(<WebexRemoteMedia meetingID="noMedia" />)).toMatchSnapshot();
    });
    test('matches snapshot of disabled remote audio', () => {
      expect(shallow(<WebexRemoteMedia meetingID="remoteVideo" />)).toMatchSnapshot();
    });

    test('matches snapshot of disabled remote video', () => {
      expect(shallow(<WebexRemoteMedia meetingID="remoteAudio" />)).toMatchSnapshot();
    });

    test('matches snapshot of enabled remote audio & video', () => {
      expect(shallow(<WebexRemoteMedia meetingID="remoteAudio&Video" />)).toMatchSnapshot();
    });

    test('matches snapshot on error', () => {
      expect(shallow(<WebexRemoteMedia meetingID="failMeetingID" />)).toMatchSnapshot();
    });
  });
});
