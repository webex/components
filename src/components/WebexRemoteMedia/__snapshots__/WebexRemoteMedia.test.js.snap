// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webex Remote Media component snapshot matches snapshot of disabled remote audio 1`] = `
<div
  className="wxc-remote-media"
>
  <video
    autoPlay={true}
    playsInline={true}
  />
</div>
`;

exports[`Webex Remote Media component snapshot matches snapshot of disabled remote video 1`] = `
<div
  className="wxc-remote-media"
>
  <audio
    autoPlay={true}
  />
</div>
`;

exports[`Webex Remote Media component snapshot matches snapshot of enabled remote audio & video 1`] = `
<div
  className="wxc-remote-media"
>
  <video
    autoPlay={true}
    playsInline={true}
  />
  <audio
    autoPlay={true}
  />
</div>
`;

exports[`Webex Remote Media component snapshot matches snapshot of enabled remote audio & video with custom CSS class 1`] = `
<div
  className="wxc-remote-media my-custom-class"
>
  <video
    autoPlay={true}
    playsInline={true}
  />
  <audio
    autoPlay={true}
  />
</div>
`;

exports[`Webex Remote Media component snapshot matches snapshot on error 1`] = `
<div
  className="wxc-remote-media"
>
  <AlertBanner
    className=""
    closable={false}
    closeBtnProps={null}
    onHide={null}
    onKeyDownClose={null}
    show={true}
    type="warning"
  >
    Having trouble joining the meeting? Please check your connection.
  </AlertBanner>
</div>
`;

exports[`Webex Remote Media component snapshot matches snapshot while loading 1`] = `
<div
  className="wxc-remote-media"
>
  <Badge
    className=""
    rounded={true}
  >
    <Spinner
      className=""
      color="black"
      percentage={null}
      showCheck={false}
      showPercentage={false}
      size={16}
    />
    <div>
      Connecting
    </div>
  </Badge>
</div>
`;
